go/test/endtoend/tabletmanager/throttler/throttler_test.go

```diff
@@ -0,0 +1,197 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package master
+
+import (
+	"flag"
+	"fmt"
+	"net/http"
+	"os"
+	"testing"
+	"time"
+
+	"vitess.io/vitess/go/vt/vttablet/tabletserver/throttle/base"
+
+	"vitess.io/vitess/go/test/endtoend/cluster"
+
+	"github.com/stretchr/testify/assert"
+)
+
+var (
+	clusterInstance *cluster.LocalProcessCluster
+	masterTablet    cluster.Vttablet
+	replicaTablet   cluster.Vttablet
+	hostname        = "localhost"
+	keyspaceName    = "ks"
+	cell            = "zone1"
+	sqlSchema       = `
+	create table t1(
+		id bigint,
+		value varchar(16),
+		primary key(id)
+	) Engine=InnoDB;
+`
+
+	vSchema = `
+	{
+    "sharded": true,
+    "vindexes": {
+      "hash": {
+        "type": "hash"
+      }
+    },
+    "tables": {
+      "t1": {
+        "column_vindexes": [
+          {
+            "column": "id",
+            "name": "hash"
+          }
+        ]
+      }
+    }
+	}`
+
+	httpClient   = base.SetupHTTPClient(time.Second)
+	checkAPIPath = "throttler/check"
+)
+
+func TestMain(m *testing.M) {
+	defer cluster.PanicHandler(nil)
+	flag.Parse()
+
+	exitCode := func() int {
+		clusterInstance = cluster.NewCluster(cell, hostname)
+		defer clusterInstance.Teardown()
+
+		// Start topo server
+		err := clusterInstance.StartTopo()
+		if err != nil {
+			return 1
+		}
+
+		// Set extra tablet args for lock timeout
+		clusterInstance.VtTabletExtraArgs = []string{
+			"-lock_tables_timeout", "5s",
+			"-watch_replication_stream",
+			"-enable_replication_reporter",
+			"-heartbeat_enable",
+			"-heartbeat_interval", "250ms",
+		}
+		// We do not need semiSync for this test case.
+		clusterInstance.EnableSemiSync = false
+
+		// Start keyspace
+		keyspace := &cluster.Keyspace{
+			Name:      keyspaceName,
+			SchemaSQL: sqlSchema,
+			VSchema:   vSchema,
+		}
+
+		if err = clusterInstance.StartUnshardedKeyspace(*keyspace, 1, false); err != nil {
+			return 1
+		}
+
+		// Collect table paths and ports
+		tablets := clusterInstance.Keyspaces[0].Shards[0].Vttablets
+		for _, tablet := range tablets {
+			if tablet.Type == "master" {
+				masterTablet = *tablet
+			} else if tablet.Type != "rdonly" {
+				replicaTablet = *tablet
+			}
+		}
+
+		return m.Run()
+	}()
+	os.Exit(exitCode)
+}
+
+func throttleCheck() (*http.Response, error) {
+	return httpClient.Head(fmt.Sprintf("http://localhost:%d/%s", masterTablet.HTTPPort, checkAPIPath))
+}
+
+func TestThrottlerBeforeMetricsCollected(t *testing.T) {
+	defer cluster.PanicHandler(t)
+
+	// Immediately after startup, we expect this response:
+	// {"StatusCode":404,"Value":0,"Threshold":0,"Message":"No such metric"}
+	resp, err := throttleCheck()
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusNotFound, resp.StatusCode)
+}
+
+func TestThrottlerAfterMetricsCollected(t *testing.T) {
```

Comment:
This test, and the next one, is doing fair amount of `time.Sleep`. Is there no way to speed these tests up a bit?

Reply:
I guess the way to speed this up is to make the throttler's roster polling interval smaller, and to implement that by adding a new configuration variable. However, does it make sense to only add this configuration variable for purposes of CI tests?

Reply:
Does it make sense for users to configure the polling interval?

Reply:
See my answer to @harshit-gangal : https://github.com/vitessio/vitess/pull/6668#discussion_r496687787

> Based on experience, these are good numbers in production, and the risk for letting the user change these is that they'll use too small intervals (server too busy, and also meaningless since we have some error margin due to heartbeat injection), or too relaxed intervals (computation not reliable enough, causing high lags). I'm in the opinion of being opinionated to a reasonable degree, and feel comfortable with these values.